File: backend/symbiont/shared_models.py
"""
Shared models configuration for consistent model management across frontend and backend.
Reads from the shared-models-config.json file at repository root.
"""
import json
import os
import re
from pathlib import Path
from typing import Dict, List, Optional, Tuple

# Get the repository root (parent of backend directory)
REPO_ROOT = Path(__file__).parent.parent.parent
CONFIG_FILE = REPO_ROOT / "shared-models-config.json"

class SharedModelsConfig:
    """Utility class to read and work with shared models configuration."""
    
    def __init__(self):
        self._config = None
        self._load_config()
    
    def _load_config(self):
        """Load the shared models configuration from JSON file."""
        try:
            with open(CONFIG_FILE, 'r', encoding='utf-8') as f:
                self._config = json.load(f)
        except FileNotFoundError:
            raise FileNotFoundError(
                f"Shared models config file not found at {CONFIG_FILE}. "
                "Make sure shared-models-config.json exists at repository root."
            )
        except json.JSONDecodeError as e:
            raise ValueError(f"Invalid JSON in shared models config: {e}")
    
    def get_all_models(self) -> List[Dict]:
        """Get all models from all providers as a flat list."""
        all_models = []
        for provider_models in self._config["models"].values():
            all_models.extend(provider_models)
        return all_models
    
    def get_models_by_provider(self, provider: str) -> List[Dict]:
        """Get all models for a specific provider."""
        return self._config["models"].get(provider, [])
    
    def get_model_by_id(self, model_id: str) -> Optional[Dict]:
        """Get a specific model by its ID."""
        for model in self.get_all_models():
            if model["id"] == model_id:
                return model
        return None
    
    def get_display_name(self, model_id: str) -> Optional[str]:
        """Get the display name for a model ID."""
        model = self.get_model_by_id(model_id)
        return model["displayName"] if model else None
    
    def detect_provider(self, model_name: str) -> Optional[str]:
        """Detect the provider for a given model name using regex patterns."""
        provider_detection = self._config.get("providerDetection", {})
        
        for provider, config in provider_detection.items():
            pattern = config.get("pattern", "")
            if pattern and re.match(pattern, model_name):
                return provider
        
        return None
    
    def is_openai_model(self, model_name: str) -> bool:
        """Check if a model is an OpenAI model."""
        return self.detect_provider(model_name) == "openai"
    
    def is_anthropic_model(self, model_name: str) -> bool:
        """Check if a model is an Anthropic model."""
        return self.detect_provider(model_name) == "anthropic"
    
    def is_google_model(self, model_name: str) -> bool:
        """Check if a model is a Google model."""
        return self.detect_provider(model_name) == "google"
    
    def is_custom_model(self, model_name: str) -> bool:
        """Check if a model is a custom model."""
        return self.detect_provider(model_name) == "custom"


# Singleton instance
_shared_models_config = None

def get_shared_models_config() -> SharedModelsConfig:
    """Get the singleton instance of SharedModelsConfig."""
    global _shared_models_config
    if _shared_models_config is None:
        _shared_models_config = SharedModelsConfig()
    return _shared_models_config

# Convenience functions for backward compatibility
def isOpenAImodel(llm_name: str) -> bool:
    """Check if a model is an OpenAI model."""
    return get_shared_models_config().is_openai_model(llm_name)

def isAnthropicModel(llm_name: str) -> bool:
    """Check if a model is an Anthropic model."""
    return get_shared_models_config().is_anthropic_model(llm_name)

def isGoogleModel(llm_name: str) -> bool:
    """Check if a model is a Google model."""
    return get_shared_models_config().is_google_model(llm_name)

def isCustomModel(llm_name: str) -> bool:
    """Check if a model is a custom model."""
    return get_shared_models_config().is_custom_model(llm_name)

File: frontend/src/utils/shared-models.ts
/**
 * Shared models configuration for consistent model management across frontend and backend.
 * Reads from the shared-models-config.json file at repository root.
 */

// Import the shared config directly
import sharedModelsConfig from '../../../shared-models-config.json';

export interface ModelConfig {
  id: string;
  displayName: string;
  provider: string;
  category: string;
}

export interface ProviderDetection {
  pattern: string;
  description: string;
}

export interface SharedModelsConfig {
  models: {
    openai: ModelConfig[];
    anthropic: ModelConfig[];
    google: ModelConfig[];
    custom: ModelConfig[];
  };
  providerDetection: {
    [provider: string]: ProviderDetection;
  };
}

class SharedModelsConfigUtil {
  private config: SharedModelsConfig;

  constructor() {
    this.config = sharedModelsConfig as SharedModelsConfig;
  }

  /**
   * Get all models from all providers as a flat list
   */
  getAllModels(): ModelConfig[] {
    const allModels: ModelConfig[] = [];
    Object.values(this.config.models).forEach(providerModels => {
      allModels.push(...providerModels);
    });
    return allModels;
  }

  /**
   * Get all models for a specific provider
   */
  getModelsByProvider(provider: keyof SharedModelsConfig['models']): ModelConfig[] {
    return this.config.models[provider] || [];
  }

  /**
   * Get a specific model by its ID
   */
  getModelById(modelId: string): ModelConfig | undefined {
    return this.getAllModels().find(model => model.id === modelId);
  }

  /**
   * Get the display name for a model ID
   */
  getDisplayName(modelId: string): string | undefined {
    const model = this.getModelById(modelId);
    return model?.displayName;
  }

  /**
   * Detect the provider for a given model name using regex patterns
   */
  detectProvider(modelName: string): string | null {
    const providerDetection = this.config.providerDetection;
    
    for (const [provider, config] of Object.entries(providerDetection)) {
      const pattern = config.pattern;
      if (pattern && new RegExp(pattern).test(modelName)) {
        return provider;
      }
    }
    
    return null;
  }

  /**
   * Check if a model is an OpenAI model
   */
  isOpenAIModel(modelName: string): boolean {
    return this.detectProvider(modelName) === 'openai';
  }

  /**
   * Check if a model is an Anthropic model
   */
  isAnthropicModel(modelName: string): boolean {
    return this.detectProvider(modelName) === 'anthropic';
  }

  /**
   * Check if a model is a Google model
   */
  isGoogleModel(modelName: string): boolean {
    return this.detectProvider(modelName) === 'google';
  }

  /**
   * Check if a model is a custom model
   */
  isCustomModel(modelName: string): boolean {
    return this.detectProvider(modelName) === 'custom';
  }

  /**
   * Get models formatted for UI dropdown/select components
   */
  getModelsForUI(): Array<{ value: string; label: string }> {
    return this.getAllModels().map(model => ({
      value: model.id,
      label: model.displayName
    }));
  }

  /**
   * Get models grouped by provider for organized UI display
   */
  getGroupedModelsForUI(): Array<{
    provider: string;
    models: Array<{ value: string; label: string }>;
  }> {
    return [
      {
        provider: 'OpenAI',
        models: this.getModelsByProvider('openai').map(m => ({ 
          value: m.id, 
          label: m.displayName 
        }))
      },
      {
        provider: 'Anthropic',
        models: this.getModelsByProvider('anthropic').map(m => ({ 
          value: m.id, 
          label: m.displayName 
        }))
      },
      {
        provider: 'Google',
        models: this.getModelsByProvider('google').map(m => ({ 
          value: m.id, 
          label: m.displayName 
        }))
      },
      {
        provider: 'Custom',
        models: this.getModelsByProvider('custom').map(m => ({ 
          value: m.id, 
          label: m.displayName 
        }))
      }
    ].filter(group => group.models.length > 0);
  }
}

// Singleton instance
let sharedModelsConfigUtil: SharedModelsConfigUtil | null = null;

export function getSharedModelsConfig(): SharedModelsConfigUtil {
  if (sharedModelsConfigUtil === null) {
    sharedModelsConfigUtil = new SharedModelsConfigUtil();
  }
  return sharedModelsConfigUtil;
}

// Create enum from shared config for backward compatibility
const models = sharedModelsConfig as SharedModelsConfig;
export const LLMModels = Object.fromEntries(
  Object.values(models.models)
    .flat()
    .map(model => [
      model.id.toUpperCase().replace(/[.-]/g, '_').replace(/\//g, '_'),
      model.id
    ])
);

// Convenience functions for backward compatibility
export function isOpenAIModel(modelName: string): boolean {
  return getSharedModelsConfig().isOpenAIModel(modelName);
}

export function isAnthropicModel(modelName: string): boolean {
  return getSharedModelsConfig().isAnthropicModel(modelName);
}

export function isGoogleModel(modelName: string): boolean {
  return getSharedModelsConfig().isGoogleModel(modelName);
}

export function isCustomModel(modelName: string): boolean {
  return getSharedModelsConfig().isCustomModel(modelName);
}